### PR TITLE
java-backend bugfix: ConjunctiveFormula.simplify()

### PR DESCRIPTION
**Pull request by denis-bogdanas**
properly restore the constraint on all execution paths.

Change description: added one big try-finally block, and restored initial constraint in finally{}.

After I implemented improved simplification in some previous PR, constraint changes during simplification process. Previously, it was not restored on all exit paths. Now it is properly restored in a finally block. Finally block is needed to minimize changes, because there are multiple return statements in the function.

DO NOT MERGE

### CODE REVIEW

File: java-backend/src/main/java/org/kframework/backend/java/symbolic/ConjunctiveFormula.java [hunk 1]
```diff
@@ -386,127 +386,130 @@ private ConjunctiveFormula simplify(boolean patternFolding, boolean partialSimpl
...
-        } while (change);
+                equalities = pendingEqualities;
+            } while (change);
 
-        context.setTopConstraint(originalTopConstraint);
-        return ConjunctiveFormula.of(substitution, equalities, disjunctions, global);
+            return ConjunctiveFormula.of(substitution, equalities, disjunctions, global);
+        } finally {
+            context.setTopConstraint(originalTopConstraint);
+        }
```
Comment (daejunpark) [inline java-backend/src/main/java/org/kframework/backend/java/symbolic/ConjunctiveFormula.java:512]: Github diff is not smart enough to show the meaningful difference here. :(

So, the main difference is the addition of the `finally` block here? Or, do you have other changes inside the `try` block?

Comment (daejunpark) [reply java-backend/src/main/java/org/kframework/backend/java/symbolic/ConjunctiveFormula.java:512]: If the `finally` block is the only change made, then it looks good to me, and you can merge.

Comment (denis-bogdanas) [reply java-backend/src/main/java/org/kframework/backend/java/symbolic/ConjunctiveFormula.java:512]: Yes, finally block is the only change.

**Review by daejunpark — APPROVED** [review]

**Comment by ehildenb** [comment]
Can one of the admins verify this patch?

**Comment by denis-bogdanas** [comment]
@ehildenb I want to run these 4 PRs on VSC first, but I'm busy with other tasks for the moment. Will get to them in 1-2 days. The 4 PRs were already reviewed by @daejunpark .

**Pull request merged**